File: script/make/protocol/lua/meta/ProtocolDefine.lua
local accountProtocol = require("./AccountProtocol")
local roleProtocol = require("./RoleProtocol")
local itemProtocol = require("./ItemProtocol")
local taskProtocol = require("./TaskProtocol")
local shopProtocol = require("./ShopProtocol")
local mailProtocol = require("./MailProtocol")
local friendProtocol = require("./FriendProtocol")
local chatProtocol = require("./ChatProtocol")
local skillProtocol = require("./SkillProtocol")
local buffProtocol = require("./BuffProtocol")
local titleProtocol = require("./TitleProtocol")
local fashionProtocol = require("./FashionProtocol")
local bubbleProtocol = require("./BubbleProtocol")
local achievementProtocol = require("./AchievementProtocol")
local dailyProtocol = require("./DailyProtocol")
local welfareProtocol = require("./WelfareProtocol")
local auctionProtocol = require("./AuctionProtocol")
local dungeonProtocol = require("./DungeonProtocol")
local warProtocol = require("./WarProtocol")
local rankProtocol = require("./RankProtocol")
local rankCenterProtocol = require("./RankCenterProtocol")
local rankWorldProtocol = require("./RankWorldProtocol")
local mapProtocol = require("./MapProtocol")
local guildProtocol = require("./GuildProtocol")
local noticeProtocol = require("./NoticeProtocol")
local cheatProtocol = require("./CheatProtocol")
local testProtocol = require("./TestProtocol")

ProtocolDefine = {}

function ProtocolDefine.getRead(protocol, type)
    local number = math.floor(protocol / 100)
    if number == 100 then 
        return accountProtocol[protocol].read
    elseif number == 101 then 
        return roleProtocol[protocol].read
    elseif number == 111 then 
        return itemProtocol[protocol].read
    elseif number == 112 then 
        return taskProtocol[protocol].read
    elseif number == 113 then 
        return shopProtocol[protocol].read
    elseif number == 114 then 
        return mailProtocol[protocol].read
    elseif number == 115 then 
        return friendProtocol[protocol].read
    elseif number == 116 then 
        return chatProtocol[protocol].read
    elseif number == 117 then 
        return skillProtocol[protocol].read
    elseif number == 118 then 
        return buffProtocol[protocol].read
    elseif number == 119 then 
        return titleProtocol[protocol].read
    elseif number == 120 then 
        return fashionProtocol[protocol].read
    elseif number == 121 then 
        return bubbleProtocol[protocol].read
    elseif number == 122 then 
        return achievementProtocol[protocol].read
    elseif number == 123 then 
        return dailyProtocol[protocol].read
    elseif number == 150 then 
        return welfareProtocol[protocol].read
    elseif number == 161 then 
        return auctionProtocol[protocol].read
    elseif number == 170 then 
        return dungeonProtocol[protocol].read
    elseif number == 180 then 
        return warProtocol[protocol].read
    elseif number == 190 then 
        return rankProtocol[protocol].read
    elseif number == 191 then 
        return rankCenterProtocol[protocol].read
    elseif number == 192 then 
        return rankWorldProtocol[protocol].read
    elseif number == 200 then 
        return mapProtocol[protocol].read
    elseif number == 301 then 
        return guildProtocol[protocol].read
    elseif number == 500 then 
        return noticeProtocol[protocol].read
    elseif number == 600 then 
        return cheatProtocol[protocol].read
    elseif number == 655 then 
        return testProtocol[protocol].read
    else
        error(string.format("unknown protocol define: %d", protocol))
    end
end

function ProtocolDefine.getWrite(protocol, type)
    local number = math.floor(protocol / 100)
    if number == 100 then 
        return accountProtocol[protocol].write
    elseif number == 101 then 
        return roleProtocol[protocol].write
    elseif number == 111 then 
        return itemProtocol[protocol].write
    elseif number == 112 then 
        return taskProtocol[protocol].write
    elseif number == 113 then 
        return shopProtocol[protocol].write
    elseif number == 114 then 
        return mailProtocol[protocol].write
    elseif number == 115 then 
        return friendProtocol[protocol].write
    elseif number == 116 then 
        return chatProtocol[protocol].write
    elseif number == 117 then 
        return skillProtocol[protocol].write
    elseif number == 118 then 
        return buffProtocol[protocol].write
    elseif number == 119 then 
        return titleProtocol[protocol].write
    elseif number == 120 then 
        return fashionProtocol[protocol].write
    elseif number == 121 then 
        return bubbleProtocol[protocol].write
    elseif number == 122 then 
        return achievementProtocol[protocol].write
    elseif number == 123 then 
        return dailyProtocol[protocol].write
    elseif number == 150 then 
        return welfareProtocol[protocol].write
    elseif number == 161 then 
        return auctionProtocol[protocol].write
    elseif number == 170 then 
        return dungeonProtocol[protocol].write
    elseif number == 180 then 
        return warProtocol[protocol].write
    elseif number == 190 then 
        return rankProtocol[protocol].write
    elseif number == 191 then 
        return rankCenterProtocol[protocol].write
    elseif number == 192 then 
        return rankWorldProtocol[protocol].write
    elseif number == 200 then 
        return mapProtocol[protocol].write
    elseif number == 301 then 
        return guildProtocol[protocol].write
    elseif number == 500 then 
        return noticeProtocol[protocol].write
    elseif number == 600 then 
        return cheatProtocol[protocol].write
    elseif number == 655 then 
        return testProtocol[protocol].write
    else
        error(string.format("unknown protocol define: %d", protocol))
    end
end

File: script/make/protocol/js/meta/ProtocolDefine.js
import accountProtocol from "./AccountProtocol.js";
import roleProtocol from "./RoleProtocol.js";
import itemProtocol from "./ItemProtocol.js";
import taskProtocol from "./TaskProtocol.js";
import shopProtocol from "./ShopProtocol.js";
import mailProtocol from "./MailProtocol.js";
import friendProtocol from "./FriendProtocol.js";
import chatProtocol from "./ChatProtocol.js";
import skillProtocol from "./SkillProtocol.js";
import buffProtocol from "./BuffProtocol.js";
import titleProtocol from "./TitleProtocol.js";
import fashionProtocol from "./FashionProtocol.js";
import bubbleProtocol from "./BubbleProtocol.js";
import achievementProtocol from "./AchievementProtocol.js";
import dailyProtocol from "./DailyProtocol.js";
import welfareProtocol from "./WelfareProtocol.js";
import auctionProtocol from "./AuctionProtocol.js";
import dungeonProtocol from "./DungeonProtocol.js";
import warProtocol from "./WarProtocol.js";
import rankProtocol from "./RankProtocol.js";
import rankCenterProtocol from "./RankCenterProtocol.js";
import rankWorldProtocol from "./RankWorldProtocol.js";
import mapProtocol from "./MapProtocol.js";
import guildProtocol from "./GuildProtocol.js";
import noticeProtocol from "./NoticeProtocol.js";
import cheatProtocol from "./CheatProtocol.js";
import testProtocol from "./TestProtocol.js";

export default class ProtocolDefine {
    static getRead(protocol) {
        switch (Math.trunc(protocol / 100)) {
            case 100: return accountProtocol[protocol].read;
            case 101: return roleProtocol[protocol].read;
            case 111: return itemProtocol[protocol].read;
            case 112: return taskProtocol[protocol].read;
            case 113: return shopProtocol[protocol].read;
            case 114: return mailProtocol[protocol].read;
            case 115: return friendProtocol[protocol].read;
            case 116: return chatProtocol[protocol].read;
            case 117: return skillProtocol[protocol].read;
            case 118: return buffProtocol[protocol].read;
            case 119: return titleProtocol[protocol].read;
            case 120: return fashionProtocol[protocol].read;
            case 121: return bubbleProtocol[protocol].read;
            case 122: return achievementProtocol[protocol].read;
            case 123: return dailyProtocol[protocol].read;
            case 150: return welfareProtocol[protocol].read;
            case 161: return auctionProtocol[protocol].read;
            case 170: return dungeonProtocol[protocol].read;
            case 180: return warProtocol[protocol].read;
            case 190: return rankProtocol[protocol].read;
            case 191: return rankCenterProtocol[protocol].read;
            case 192: return rankWorldProtocol[protocol].read;
            case 200: return mapProtocol[protocol].read;
            case 301: return guildProtocol[protocol].read;
            case 500: return noticeProtocol[protocol].read;
            case 600: return cheatProtocol[protocol].read;
            case 655: return testProtocol[protocol].read;
            default:throw("unknown protocol define: " + protocol)
        }
    }

    static getWrite(protocol) {
        switch (Math.trunc(protocol / 100)) {
            case 100: return accountProtocol[protocol].write;
            case 101: return roleProtocol[protocol].write;
            case 111: return itemProtocol[protocol].write;
            case 112: return taskProtocol[protocol].write;
            case 113: return shopProtocol[protocol].write;
            case 114: return mailProtocol[protocol].write;
            case 115: return friendProtocol[protocol].write;
            case 116: return chatProtocol[protocol].write;
            case 117: return skillProtocol[protocol].write;
            case 118: return buffProtocol[protocol].write;
            case 119: return titleProtocol[protocol].write;
            case 120: return fashionProtocol[protocol].write;
            case 121: return bubbleProtocol[protocol].write;
            case 122: return achievementProtocol[protocol].write;
            case 123: return dailyProtocol[protocol].write;
            case 150: return welfareProtocol[protocol].write;
            case 161: return auctionProtocol[protocol].write;
            case 170: return dungeonProtocol[protocol].write;
            case 180: return warProtocol[protocol].write;
            case 190: return rankProtocol[protocol].write;
            case 191: return rankCenterProtocol[protocol].write;
            case 192: return rankWorldProtocol[protocol].write;
            case 200: return mapProtocol[protocol].write;
            case 301: return guildProtocol[protocol].write;
            case 500: return noticeProtocol[protocol].write;
            case 600: return cheatProtocol[protocol].write;
            case 655: return testProtocol[protocol].write;
            default:throw("unknown protocol define: " + protocol)
        }
    }
}